contracts/upgradeable_contracts/U_ForeignBridge.sol

```diff
@@ -150,34 +152,37 @@ contract ForeignBridge is ERC677Receiver, Validatable {
     function submitSignature(bytes signature, bytes message) public onlyValidator {
         // ensure that `signature` is really `message` signed by `msg.sender`
         require(msg.sender == MessageSigning.recoverAddressFromSignedMessage(signature, message));
+        bytes32 hashMsg = keccak256(message);
+        bytes32 hashSender = keccak256(msg.sender, hashMsg);
 
-        require(message.length == 116);
-        bytes32 hash = keccak256(message);
-        bytes32 hashSender = keccak256(msg.sender, hash);
-
-        uint256 signed = numMessagesSigned(hash) + 1;
-
+        uint256 signed = numMessagesSigned(hashMsg);
+        require(!isDepositProcessed(signed));
+        signed = signed.add(1);
         if (signed > 1) {
             // Duplicated signatures
             require(!messagesSigned(hashSender));
         } else {
-            // check if it will really reduce gas usage in case of the second transaction
-            // with the same hash
-            setMessages(hash, message);
+            setMessages(hashMsg, message);
         }
         setMessagesSigned(hashSender, true);
 
-        bytes32 signIdx = keccak256(hash, (signed-1));
+        bytes32 signIdx = keccak256(hashMsg, (signed-1));
         setSignatures(signIdx, signature);
 
-        setNumMessagesSigned(hash, signed);
+        setNumMessagesSigned(hashMsg, signed);
 
-        emit SignedForWithdraw(msg.sender, hash);
+        emit SignedForWithdraw(msg.sender, hashMsg);
         if (signed == validatorContract().requiredSignatures()) {
-            emit CollectedSignatures(msg.sender, hash);
+            uint256 protectionNumber = signed | 2**255;
+            setNumMessagesSigned(hashMsg, protectionNumber);
+            emit CollectedSignatures(msg.sender, hashMsg);
         }
     }
 
+    function isDepositProcessed(uint256 _number) public view returns(bool) {
```

Comment:
since the method is called for two different scenarios: depost and withdraw, let's give a common name to it like `isAlreadyProcessed`